src/Engine.zig

```diff
@@ -63,19 +67,41 @@ delta_time_ns: u64 = 0,
 timer: std.time.Timer,
 
 pub const Core = struct {
-    internal: union {
-        window: glfw.Window,
-    },
+    internal: GetCoreInternalType(),
 };
 
 pub const GpuDriver = struct {
+    internal: GetGpuDriverInternalType(),
+
     device: gpu.Device,
     backend_type: gpu.Adapter.BackendType,
     swap_chain: ?gpu.SwapChain,
     swap_chain_format: gpu.Texture.Format,
 
-    native_instance: gpu.NativeInstance,
     surface: ?gpu.Surface,
     current_desc: gpu.SwapChain.Descriptor,
     target_desc: gpu.SwapChain.Descriptor,
 };
+
+pub fn init(allocator: std.mem.Allocator, options: Options) !Engine {
+    var engine = Engine{
+        .allocator = allocator,
+        .options = options,
+        .timer = try std.time.Timer.start(),
+        .core = undefined,
+        .gpu_driver = undefined,
+    };
+
+    engine.core.internal = try GetCoreInternalType().init(allocator, &engine);
+    engine.gpu_driver.internal = try GetGpuDriverInternalType().init(allocator, &engine);
+
+    return engine;
+}
+
+fn GetCoreInternalType() type {
+    return @import("native.zig").CoreGlfw;
+}
+
+fn GetGpuDriverInternalType() type {
+    return @import("native.zig").GpuDriverNative;
```

Comment:
I haven't looked closely enough yet to see how viable this is, but here's what I am thinking:

* Maybe instead of two separate `Core` and `GpuDriver` types, we could have just one `Platform` type (which handles platform specifics)
* Maybe we should not use the name `internal` (implies "never touch this") and instead use `platform` (it would be possible to expose platform-specific functionality behind these, like say ability to send a push notification on android/ios as an example - although I don't know we actually would want such a feature.)

Just a thought I'll explore later, sharing in case any of it resonates with you. Happy to continue down this code path / design for now 👍

Reply:
1) Yeah that would work well especially because Core and GpuDriver are quite related 
2) Agreed